py/obj.h

```diff
@@ -743,15 +743,30 @@ void *mp_obj_malloc_helper(size_t num_bytes, const mp_obj_type_t *type);
 // check for more specific object types.
 // Note: these are kept as macros because inline functions sometimes use much
 // more code space than the equivalent macros, depending on the compiler.
-#define mp_obj_is_type(o, t) (mp_obj_is_obj(o) && (((mp_obj_base_t *)MP_OBJ_TO_PTR(o))->type == (t))) // this does not work for checking int, str or fun; use below macros for that
+// don't use mp_obj_is_exact_type directly; use mp_obj_is_type which provides additional safety checks.
+// use the former only if you need to bypass these checks (because you've already checked everything else)
+#define mp_obj_is_exact_type(o, t) (mp_obj_is_obj(o) && (((mp_obj_base_t *)MP_OBJ_TO_PTR(o))->type == (t))) // this does not work for checking int, str or fun; use below macros for that
+
+// type checks are split to a separate, constant result macro. I found this way to work without hindering GCC's
+// optimizations (other tricks like using ({ expr; exper; }) or (exp, expr, expr) in mp_obj_is_type result in
+// missed optimizations. however, this way GCC retains the optimizations it used to have before these checks
+// were added).
+#define _type_checks(t) (                                                         \
```

Comment:
I'd suggest calling this something like `mp_obj_is_type_assert_helper()`.

Reply:
Neither names indicate the specific checks this macro does (str/int/...) so I preferred the obviously-private name (not beginning with `mp_`). I figure the `mp_obj_is_type_` prefix is from `mp_obj_is_type()`? That makes some sense, yes. I'll make this change

Reply:
Names beginning with `_` are reserved for the compiler, so we can't do that.  Also it's a #define in a public header so needs to be pretty unique so it doesn't clash with anything and give strange compiler errors.

Yes, my idea to prefix with `mp_obj_is_type_` is to emphasise it's a helper for `mp_obj_is_type()`.

Alternatively could call it the very explicit `mp_assert_not_type_bool_int_str_nonetype`.

Reply:
Or maybe `mp_type_assert_not_bool_int_str_nonetype`.

Reply:
I liked the explictness, going with `mp_type_assert_not_bool_int_str_nonetype`.

Reply:
Great!